js/modules/k6/http/request.go

```diff
@@ -357,70 +356,98 @@ func (h *HTTP) parseRequest(
+	results := make(map[string]*Response, reqCount)
+
+	i := 0
+	for key, req := range requests {
+		parsedReq, err := h.parseBatchRequest(ctx, key, req)
+		if err != nil {
+			return nil, nil, err
+		}
+		response := new(httpext.Response)
+		batchReqs[i] = httpext.BatchParsedHTTPRequest{
+			ParsedHTTPRequest: parsedReq,
+			Response:          response,
+		}
+		results[key] = &Response{response}
+		i++
+	}
+
+	return batchReqs, results, nil
+}
+
 // Batch makes multiple simultaneous HTTP requests. The provideds reqsV should be an array of request
 // objects. Batch returns an array of responses and/or error
 func (h *HTTP) Batch(ctx context.Context, reqsV goja.Value) (goja.Value, error) {
 	state := lib.GetState(ctx)
 	if state == nil {
 		return nil, ErrBatchForbiddenInInitContext
 	}
-	rt := common.GetRuntime(ctx)
-
-	reqs := reqsV.ToObject(rt)
-	keys := reqs.Keys()
-	parsedReqs := map[string]*httpext.ParsedHTTPRequest{}
-	for _, key := range keys {
-		parsedReq, err := h.parseBatchRequest(ctx, key, reqs.Get(key))
-		if err != nil {
-			return nil, err
-		}
-		parsedReqs[key] = parsedReq
-	}
 
 	var (
-		// Return values; retval must be guarded by the mutex.
-		mutex  sync.Mutex
-		retval = rt.NewObject()
-		errs   = make(chan error)
-
-		// Concurrency limits.
-		globalLimiter  = NewSlotLimiter(int(state.Options.Batch.Int64))
-		perHostLimiter = NewMultiSlotLimiter(int(state.Options.BatchPerHost.Int64))
+		err       error
+		batchReqs []httpext.BatchParsedHTTPRequest
+		results   interface{} // either []*Response or map[string]*Response
 	)
-	for k, pr := range parsedReqs {
-		go func(key string, parsedReq *httpext.ParsedHTTPRequest) {
-			globalLimiter.Begin()
-			defer globalLimiter.End()
-
-			if hl := perHostLimiter.Slot(parsedReq.URL.GetURL().Host); hl != nil {
-				hl.Begin()
-				defer hl.End()
-			}
-
-			res, err := httpext.MakeRequest(ctx, parsedReq)
-			if err != nil {
-				errs <- err
-				return
-			}
 
-			mutex.Lock()
-			_ = retval.Set(key, responseFromHttpext(res))
-			mutex.Unlock()
+	switch v := reqsV.Export().(type) {
+	case []interface{}:
+		batchReqs, results, err = h.prepareBatchArray(ctx, v)
+	case map[string]interface{}:
+		batchReqs, results, err = h.prepareBatchObject(ctx, v)
+	default:
+		return nil, fmt.Errorf("invalid http.batch() argument type %T", v)
+	}
 
-			errs <- nil
-		}(k, pr)
+	if err != nil {
+		return nil, err
 	}
 
-	var err error
-	for range keys {
+	reqCount := len(batchReqs)
+	errs := httpext.MakeBatchRequests(
+		ctx, batchReqs, reqCount,
+		int(state.Options.Batch.Int64), int(state.Options.BatchPerHost.Int64),
+	)
+
+	for i := 0; i < reqCount; i++ {
 		if e := <-errs; e != nil {
 			err = e
 		}
 	}
-	return retval, err
+	return common.GetRuntime(ctx).ToValue(results), err
 }
 
-func (h *HTTP) parseBatchRequest(ctx context.Context, key string, val goja.Value) (*httpext.ParsedHTTPRequest, error) {
+func (h *HTTP) parseBatchRequest(
+	ctx context.Context, key interface{}, val interface{},
```

Comment:
Can we do this without `interface{}`?

If not, please add documentation. Looking at the `interface{}` type gains nothing for the reader.

Reply:
if you figure out how we can do it without interface, let me know - half the purpose of this method is to find out the type of this value :wink: